source/geometry/Quaternion.ooc

```diff
@@ -234,7 +234,12 @@ Quaternion: cover {
 		// W - observationVectors
 		// Y - gibbsVector
 		// Z - vectorQuantityZ
-		// TODO: Fix singularity problem when the angle is close to PI, using sequential rotations
+		// Rotate quaternions if the angle is close to pi to avoid singularity problem
+		centerQuaternion := quaternions[quaternions count / 2]
+		preRotate := centerQuaternion angle(This identity) equals(Float pi, 0.5f)
+		if (preRotate)
+			for (index in 0 .. quaternions count)
+				quaternions[index] = centerQuaternion inverse * quaternions[index]
```

Comment:
You calculate the inverse on every iteration now. Wouldn't it be better to do it just once?